gpsgen.py

```diff
@@ -1,9 +1,516 @@
+        f_string = "[<STEP> for {}] [<ACTION>:{}] [<COORD>:{}] [<TIME>: {}]"
+        return f_string.format(*step_data)
+
+
+class StepScheduler():
+    def __init__(self):
+        self._step_queue = deque()
+
+    def new_step(self, step):
+        self._step_queue.append(step)
+
+    def run(self):
+        while self._step_queue:
+            step = self._step_queue.popleft()
+            try:
+                next(step)
+                self._step_queue.append(step)
+            except StopIteration:
+                pass
+
+def io_coroutine(gen):
+    """
+    This decorator, pumps the coroutine to the next step.
+    It is also a Syntactic sugar for identifying coroutines.
+    Which are much more visible now.
+    Registers the initiliced coroutine for closing them when
+    the data flow ends, this is specially important, when
+    opened file descriptors are involved, in this case we are
+    sending data to a "with open" block, registering coroutines
+    allow us to kill them easyly with Simulator.end_io_coroutines()
+    """
+    def pumped(self, *args, **kwargs):
+        g = gen(self, *args, **kwargs)
+        self.io_coroutines.append(g)
+        next(g)
+        return g
+    return pumped
+
+class Simulation():
+    """
+    This is the main class of this script. It configures the simulation, starts
+    the necesary coroutines, and starts the scheduler. In the schedulers queue
+    are stored the main generator(freezed), generator that triggers each
+    coroutine when required. Add More ...
+    """
+    def __init__(self, sim_type, nworkers, hours_shift, speed, transmit_rate,
+                 is_bulk, is_json, is_pretty, start_date=None, database=None):
+        self.sim_type = sim_type
+        self.nworkers = nworkers
+        self.hours_shift = hours_shift
+        self.speed = speed
+        self.transmit_rate = transmit_rate
+        self.is_bulk = is_bulk
+        self.is_json = is_json
+        self.is_pretty = is_pretty
+        self.start_date = start_date
+        self.ends = None
+        self.database = database
+        self.scheduler = StepScheduler()
+        self.io_coroutines = []
+        self.instantiate_coroutines()
+
+    def instantiate_coroutines(self):
+        if self.is_json_file:
+            self.filewriter = self.file_writer_to_json()
+        else:
+            self.filewriter = self.file_writer()
+        self.dbwriter = self.send_to_db()
+
+    def end_io_coroutines(self):
+        for coro in self.io_coroutines:
+            try:
+                coro.send(None)
+            except StopIteration:
+                pass
+
+    @property
+    def rate(self):
+        return datetime.timedelta(seconds=self.transmit_rate)
+
+    @property
+    def starts(self):
+        if self.start_date is None:
+            return datetime.datetime.now()
+        else:
+            return self.start_date
+
+    @property
+    def is_json_file(self):
+        return any([self.is_json, self.is_pretty])
```

Comment:
My fault, i wanted to show-off. 
Your solution is much more readable.
;), changed! and I am sure it is faster to evaluate.